### PR TITLE
feat: add sheet navigation

### PR DESCRIPTION
**Pull request by quanho**
This PR is to add sheet navigation capability to support using sn-nav-menu in a mashup using qlik-embed to embed a sheet. The main idea is to create a navigation api for each sheet rendering to avoid the case when there are more than one sheet is rendered.
The navigation api needs to know the sheet to update the sheet based on new sheet ID.
The navigation api is passed to each chart context.

https://github.com/user-attachments/assets/e29414c0-3cba-48e5-86f1-6bc9d261c5d9

## Requirements checklist

- [ ] Api specification
  - [ ] Ran `yarn spec`
    - [ ] No changes **_OR_** API changes has been formally approved
- [ ] Unit/Component test coverage
- [ ] Correct PR title for the changes (fix, chore, feat)

When build and tests have passed:

- [ ] Add code reviewers, for example @qlik-oss/nebula-core

### CODE REVIEW

File: apis/nucleus/src/object/initiate-sheet.js [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
+import createNavigationApi from './navigation/navigation';
 
-export default async function initSheet(model, optional, halo, initialError, onDestroy = async () => {}) {
+export default async function initSheet(model, optional, halo, store, initialError, onDestroy = async () => {}) {
   const api = sheetAPI({
     model,
     halo,
     initialError,
     onDestroy,
   });
```
Comment (Caele) [inline apis/nucleus/src/object/initiate-sheet.js:11]: I think it looks better to pass `navigation` in here, and not into the `mount()` call. We don't want a new navigation API instance for each mount call?

Comment (quanho) [reply apis/nucleus/src/object/initiate-sheet.js]: Good idea! Fixed! Thanks!

File: apis/nucleus/src/components/sheetGlue.jsx [hunk 2]
```diff
@@ -14,10 +23,12 @@ export default function glue({ halo, element, model, initialSnOptions, initialSn
...
       onMount={onMount}
+      navigation={navigation}
     />,
     element,
     model.id
   );
+  navigation.setSheetRef(sheetRef);
 
   const unmount = () => {
     root.remove(portal);
```
Comment (Caele) [inline apis/nucleus/src/components/sheetGlue.jsx:34]: Below here is a `model.on('closed', unmount)`, won't that potentially mess with things?

**Comment by Caele** [comment]
Make sure you test this with a multi-sheet mashup, so that they don't interfere with eachother. They should be able to display the same sheet for example, and independently switch as well.

File: apis/nucleus/src/object/initiate-sheet.js (hunk 1 above)
Comment (quanho) [reply apis/nucleus/src/object/initiate-sheet.js]: To support for single nav-menu object case, the navigation creation is moved up to apis/nucleus/src/object/get-generic-object.js and from that pass navigation to either a sheet or a single nav-menu object.

**Comment by quanho** [comment]
> Make sure you test this with a multi-sheet mashup, so that they don't interfere with eachother. They should be able to display the same sheet for example, and independently switch as well.

Tested rendering the same sheet twice in a mashup using qlik-embed
https://github.com/user-attachments/assets/9f35bd50-2c21-4cae-9162-e59d1ba3f375

File: apis/supernova/src/hooks.js [hunk 3]
```diff
@@ -772,6 +772,31 @@ export function useDeviceType() {
...
+ * // Render a sheet
+ * const nebbie = embed(app);
+ * const sheetRender = await nebbie.render({
+ *   element: sheetElement
+ *   id: 'sheetId',
+ * });
+ *
+ * @example
+ * // Use a navigation object to control sheet navigation
+ * const plugins = usePlugins();
```
Comment (T-Wizard) [inline apis/supernova/src/hooks.js:789]: why usePlugin? (remove)

File: apis/supernova/src/hooks.js [hunk 4]
```diff
@@ -772,6 +772,31 @@ export function useDeviceType() {
...
+ * const navigationRender = await nebbie.render({
+ *   element: navigationElement
+ *   id: 'navigationId',
+ *   navigation: sheetRender.navigation
+ * });
```
Comment (T-Wizard) [inline apis/supernova/src/hooks.js:794]: maybe wrong place for this example?
it if for how to use navigation with .render function not how to use useNavigation

File: apis/supernova/src/hooks.js [hunk 5]
```diff
@@ -772,6 +772,31 @@ export function useDeviceType() {
...
+ */
+export function useNavigation() {
+  return useInternalContext('navigation');
+}
```
Comment (T-Wizard) [inline apis/supernova/src/hooks.js:798]: currently useNavigation always return a navigation object even if it is not connected to anything.
In my (current) opinion it would be better to return null/undefined when there is no navigation context making it the object aware if navigation is possible

**Comment by T-Wizard** [comment]
should the new exposed api be merked as "@experimental"?

File: apis/nucleus/src/object/navigation/navigation.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    setCurrentSheetId: (sheetId) => {
+      State.sheetId = sheetId;
+    },
+    /**
+     * Set the sheet ref
+     * @param {object} sheetRef sheet ref object
+     */
+    setSheetRef: (sheetRef) => {
+      State.sheetRef = sheetRef;
+    },
```
Comment (T-Wizard) [inline apis/nucleus/src/object/navigation/navigation.js:41]: I think setCurrentSheetId and setSheetRef are only intended for internal use.
maybe mark as `@private`

File: apis/nucleus/src/object/navigation/navigation.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      const { modelStore, rpcRequestModelStore } = store;
+      const key = `${sheetId}`;
+      let rpc = rpcRequestModelStore.get(key);
+      if (!rpc) {
+        rpc = halo.app.getObject(sheetId);
+        rpcRequestModelStore.set(key, rpc);
+      }
+      const model = await rpc;
+      modelStore.set(key, model);
+      State.sheetRef?.current?.setModel?.(model);
```
Comment (T-Wizard) [inline apis/nucleus/src/object/navigation/navigation.js:26]: maybe add code for if sheetId is not a valid sheet id (not an object, or an object that is not a sheet)
(here or maybe in Sheet.jsx)

File: apis/supernova/src/hooks.js (hunk 3 above)
Comment (quanho) [reply apis/supernova/src/hooks.js]: Thanks, stupid mistake from copy/paste. Fixed!

File: apis/supernova/src/hooks.js (hunk 4 above)
Comment (quanho) [reply apis/supernova/src/hooks.js]: Indeed it is a wrong place. Updated!

File: apis/supernova/src/hooks.js (hunk 5 above)
Comment (quanho) [reply apis/supernova/src/hooks.js:787]: Thanks! Fixed!

File: apis/nucleus/src/object/navigation/navigation.js (hunk 6 above)
Comment (quanho) [reply apis/nucleus/src/object/navigation/navigation.js]: Good idea! Fixed!

File: apis/nucleus/src/object/navigation/navigation.js (hunk 7 above)
Comment (quanho) [reply apis/nucleus/src/object/navigation/navigation.js]: Fixed!

File: apis/nucleus/src/object/create-session-object.js [hunk 8]
```diff
@@ -39,14 +39,15 @@ import init from './initiate';
...
-  { type, version, fields, properties, options, plugins, element, extendProperties },
+  { type, version, fields, properties, options, plugins, element, extendProperties, navigation: inputNavigation },
   halo,
   store
 ) {
   let mergedProps = {};
   const children = [];
   const { modelStore, subscribe } = store;
   let error;
+  const navigation = inputNavigation;
```
Comment (veinfors) [inline apis/nucleus/src/object/create-session-object.js:50]: No need to bring it in as `inputNavigation` here and assign it to `navigation` before returning. Might as well just take in `navigation` and return the same, if I'm not missing anything...

Comment (veinfors) [reply apis/nucleus/src/object/create-session-object.js:50]: Or, actually, I don't see that we pass in navigation into this fn at all, so probably can remove it from here completely?

File: apis/nucleus/src/components/Sheet.jsx [hunk 9]
```diff
@@ -17,10 +17,13 @@ const SheetElement = {
...
 };
 
-function getCellRenderer(cell, halo, initialSnOptions, initialSnPlugins, initialError, onMount) {
+function getCellRenderer(cell, halo, initialSnOptions, initialSnPlugins, initialError, onMount, navigation) {
   const { x, y, width, height } = cell.bounds;
   return (
-    <div style={{ left: `${x}%`, top: `${y}%`, width: `${width}%`, height: `${height}%`, position: 'absolute' }}>
+    <div
+      style={{ left: `${x}%`, top: `${y}%`, width: `${width}%`, height: `${height}%`, position: 'absolute' }}
+      key={cell.model.id}
```
Comment (Caele) [inline apis/nucleus/src/components/Sheet.jsx:25]: Whats with the key here?

File: apis/nucleus/src/components/Sheet.jsx [hunk 10]
```diff
@@ -47,115 +51,141 @@ function getBounds(pos, columns, rows) {
...
+  ) => {
+    const { root } = halo;
+    const [model, setModel] = useState(inputModel);
+    const [layout] = useLayout(model);
+    const { theme: themeName, modelStore } = useContext(InstanceContext);
+    const [cells, setCells] = useState([]);
+    const [bgColor, setBgColor] = useState(undefined);
+    const [bgImage, setBgImage] = useState(undefined);
+    const [deepHash, setDeepHash] = useState('');
+    navigation?.setCurrentSheetId?.(model.id);
```
Comment (Caele) [inline apis/nucleus/src/components/Sheet.jsx:67]: As we want some kind of event off this we need to make sure it doesn't trigger for each render.

File: apis/nucleus/src/components/glue.jsx [hunk 11]
```diff
@@ -27,6 +28,7 @@ export default function glue({
       initialError={initialError}
       onMount={onMount}
       emitter={emitter}
+      navigation={navigation}
     />,
     element,
     currentId
```
Comment (Caele) [inline apis/nucleus/src/components/glue.jsx:34]: You still have the model.on('closed', unmount) below here.

File: apis/nucleus/src/components/Sheet.jsx [hunk 12]
```diff
@@ -47,115 +51,141 @@ function getBounds(pos, columns, rows) {
...
+      const onModelClose = () => {
+        model.removeListener('closed', onModelClose);
+        unmount();
       };
-      fetchObjects();
-    }
-  }, [layout]);
+      model.on('closed', onModelClose);
+      return () => model.removeListener('closed', onModelClose);
+    }, [model]);
```
Comment (Caele) [inline apis/nucleus/src/components/Sheet.jsx:130]: Why did you move this in from sheetGlue?

File: apis/nucleus/src/components/Sheet.jsx (hunk 9 above)
Comment (quanho) [reply apis/nucleus/src/components/Sheet.jsx:25]: This is a list of cells, without a key it will use index and this causes a problem when we switch from a sheet to another sheet. Cells in a new list are not rendered, instead cells in old list are still used.

File: apis/nucleus/src/index.js [hunk 13]
```diff
@@ -84,6 +84,7 @@ const DEFAULT_SNAPSHOT_CONFIG = /** @lends SnapshotConfiguration */ {
  * @property {Array<TypeInfo>=} types Visualization types to register
  * @property {Array<ThemeInfo>=} themes Themes to register
  * @property {object=} anything
+ * @property {object=} flags
```
Comment (Caele) [inline apis/nucleus/src/index.js:87]: I don't think we want to publicly export flags like this, its an internal thing.

File: apis/nucleus/src/object/navigation/navigation.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    },
+    /**
+     * @private
+     * Set the current sheet id
+     * @param {string} sheetId Id of the current sheet
+     */
+    setCurrentSheetId: (sheetId) => {
+      State.sheetId = sheetId;
+    },
```
Comment (Caele) [inline apis/nucleus/src/object/navigation/navigation.js:46]: We want some kind of trigger/event from this right?

File: apis/nucleus/src/object/navigation/navigation.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
+export default function createNavigationApi(halo, store) {
+  const { galaxy } = halo.public;
+  if (galaxy.anything?.sense?.navigation) {
+    return galaxy.anything?.sense?.navigation;
+  }
```
Comment (T-Wizard) [inline apis/nucleus/src/object/navigation/navigation.js:5]: is this still needed?

File: apis/nucleus/src/index.js (hunk 13 above)
Comment (quanho) [reply apis/nucleus/src/index.js:87]: This is needed to pass apis?.features from qlik-embed to nebula.

Comment (Caele) [reply apis/nucleus/src/index.js:87]: no, you can override that right? Add a ts-ignore thing.

File: apis/nucleus/src/components/Sheet.jsx (hunk 10 above)
Comment (quanho) [reply apis/nucleus/src/components/Sheet.jsx:67]: There is a check if sheetId is changed or not in setCurrentSheetId, so it will do nothing if the sheetId is not changed.

File: apis/nucleus/src/object/navigation/navigation.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+      try {
+        model = await rpc;
+        if (model.genericType !== 'sheet') {
+          return;
+        }
+      } catch (e) {
+        return;
+      }
+      modelStore.set(key, model);
+      State.sheetRef?.current?.setModel?.(model);
```
Comment (Caele) [inline apis/nucleus/src/object/navigation/navigation.js:35]: I think that this should also set the current sheet ID.
Is there really any reason for ever having to call setCurrentSheetId "manually"?

File: apis/nucleus/src/object/navigation/navigation.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+     * @private
+     * Set the current sheet id
+     * @param {string} sheetId Id of the current sheet
+     */
+    setCurrentSheetId: (sheetId) => {
+      if (State.sheetId === sheetId) {
+        return;
+      }
+      State.sheetId = sheetId;
+      navigationAPI.emit('currentSheetIdChange');
```
Comment (Caele) [inline apis/nucleus/src/object/navigation/navigation.js:47]: think we should use a simpler name, like "sheetChanged", there is not much reason to have an ID change if the sheet itself didn't change.

File: apis/nucleus/src/sheet.js [hunk 18]
```diff
@@ -72,6 +72,11 @@ export default function sheet({ model, halo, initialError, onDestroy = async ()
      * @type {string}
      */
     model,
+    /**
+     * The navigation api to control sheet navigation.
+     * @type {Navigation}
+     */
```
Comment (Caele) [inline apis/nucleus/src/sheet.js:78]: add `@since vx.y.z` and `@experimental`, same for the hook.

File: apis/nucleus/src/components/Sheet.jsx [hunk 19]
```diff
@@ -47,115 +51,141 @@ function getBounds(pos, columns, rows) {
...
```
Comment (Caele) [inline apis/nucleus/src/components/Sheet.jsx:131]: One last thing, I would like to move this code out into sheet.js. As that is the "object" that owns this component and should be responsible for unmounting. However that requires a bit of restructuring as the new model is never passed through that sheet API, like changing the navigations setSheetRef to setSheetAPI and add a set model on the sheet API (could be useful for other purposes actually).
So not gonna block you with this, just consider it. I think the current setup works just fine, and is all internal anyways.

**Review by Caele — APPROVED** [review]
Think we are good! Had one last comment, but nothing we have to do now.
